Support quantified comparison predicates

AlekseiS opened the pull request.

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/ExpressionFormatter.java:
```diff
@@ -596,6 +597,20 @@ public String visitFrameBound(FrameBound node, Boolean unmangleNames)
+        @Override
+        protected String visitQuantifiedComparisonExpression(QuantifiedComparisonExpression node, Boolean unmangleNames)
+        {
+            StringBuilder result = new StringBuilder();
```
Do this with chaining

electrum commented:
Add all the parser related changes in the same commit: "Add parsing for quantified comparison predicates"

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/QuantifiedComparisonExpression.java:
```diff
@@ -0,0 +1,108 @@
+    private QuantifiedComparisonExpression(Optional<NodeLocation> location, ComparisonExpression.Type comparisonType, Quantifier quantifier, Expression left, Expression subquery)
+    {
+        super(location);
+        requireNonNull(comparisonType, "comparisonType is null");
```
Move these into the assignments:
```java
this.left = requireNonNull(left, "left is null");

electrum commented:
Add a few smoke tests for the new syntax to `TestStatementBuilder.testStatementBuilder()`

AlekseiS commented:
@martint ,

It's very error prone to add an additional parameter to ComparisonExpression. ComparisonExpression is used in numerous places including operations like normalization, boolean logic transformations, negation, optimization, etc. Most of them don't make sense for quantified comparison. So, even asserting the current behavior in all involved places is difficult and error prone. I think adding a new AST node is a much cleaner and safer approach.

Moreover, quantified comparison is logically different from normal two operand comparison as we know it. You cannot change the order of operands, you cannot negate it, one of the sides has multiple values, etc. In a way, it's more similar to "IN" operator than to comparison. It is one more reason why it's better represented by a different AST node.

Regarding doing it in planner vs optimizer, I understand the idea to do rewrites in optimizer and looked at the example you mentioned. The reasons I did it in planner are 1. because "exists" is similar and it does rewrite in the planner; 2. I'd need a new PlanNode in order to define rewrites on it if I wanted to do it in optimizer.

martint commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -267,6 +293,113 @@ private Plan plan(String sql, LogicalPlanner.Stage stage)
         }
     }
 
+    private static void assertEqualPlans(Plan actual, Plan expected)
```
There are already some utilities for making assertions over plan structures. Take a look at `com.facebook.presto.sql.planner.assertions.PlanAssert`.

AlekseiS replied:
I use this one as well. It allows to compare the resulting tree with some pre-defined pattern. In my case I also want  to compare the plan agains another query's plan which should be identical when created and optimized.

martint replied:
Those tests are brittle. There's no guarantee that the two queries are going to produce exactly the same plan. There are too many factors involved. It's reasonable to check that the queries produce the same results, and you can do this by adding entries in AbstractTestQueries.

Also, we don't want to be carrying around and maintaining yet another framework for comparing plans. Please remove them.

martint replied:
That's assuming the hand-written query gets planned in the same way, which is not necessarily true.

AlekseiS commented:
As per discussion with Martin I kept the design as it is, because there would be major complications if the plan is generated in an optimizer.
Addressed all other issues and rebased.
@martint and @electrum please have a look

martint commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/QuantifiedComparisonExpression.java:
```diff
@@ -0,0 +1,103 @@
+        return quantifier;
+    }
+
+    public Expression getLeft()
```
Call this getValue() or getExpression(). "left" is weird, since there's no corresponding "right".

AlekseiS replied:
Renamed to "value"

martint commented on presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:
```diff
@@ -1007,6 +1008,16 @@ public void testInValidJoinOnClause()
+        analyze("SELECT * FROM t1 WHERE t1.a <= ALL (VALUES 10, 20)");
+        assertFails(MULTIPLE_FIELDS_FROM_SUBQUERY, "SELECT * FROM t1 WHERE t1.a = ANY (SELECT 1, 2)");
+        assertFails(TYPE_MISMATCH, "SELECT * FROM t1 WHERE t1.a = SOME (VALUES ('abc'))");
+        assertFails(MISMATCHED_SET_COLUMN_TYPES, "SELECT * FROM t1 WHERE t1.a = ALL (VALUES ROW(10), ROW('abc'))");
```
ROW constructors are unnecessary. You can just do `VALUES 10, 'abc'`.

BTW, this might not be testing the type analysis for quantified comparison at all -- it may be failing in the analysis of VALUES.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -946,6 +951,21 @@ protected Type visitExists(ExistsPredicate node, StackableAstVisitorContext<Void
         }
 
+        @Override
+        protected Type visitQuantifiedComparisonExpression(QuantifiedComparisonExpression node, StackableAstVisitorContext<Void> context)
```
This should verify that the type of the expression implements the expected operators to be able to do the comparisons (orderable vs comparable)

AlekseiS replied:
Added a check if a type is orderable/comparable. Also, added cases to a unit test for it.
Note, that currently the change only supports orderable types for "=ALL" and "<>ANY". This check is made in SubqueryPlanner as it's not a general limitation, but is implementation specific.

AlekseiS commented:
@martint , addressed the comments.

kokosing commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/QuantifiedComparisonExpression.java:
```diff
@@ -0,0 +1,103 @@
+    {
+        ALL,
+        ANY,
+        SOME,
```
Is there any semantic difference between `ANY` and `SOME`? If there is not, the I would have only one of them (`ANY`) and just map in `AstBuilder` `SOME` to `ANY`.

AlekseiS replied:
I thought about it. One of the reason I kept both is because when formatted the query won't match the one that user gave which might be confusing.

kokosing replied:
ok, now I understand, thanks

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -384,6 +384,48 @@ public void testQuantifiedComparisonGreaterEqualAny()
+        assertPlan(query, LogicalPlanner.Stage.CREATED,
+                anyTree(
+                        project(
+                                filter("min = max AND orderkey = min",
```
`min`, `max` and `orderkey` are just aliases here. They have nothing the plan symbols. In order to check their origin you would need to write something like that.

```
        assertPlan(query, LogicalPlanner.Stage.CREATED,
                anyTree(
                        project(
                                filter("NOT (MIN = MAX AND O = MIN)",
                                        apply(ImmutableList.of(),
                                                project(
                                                        tableScan("orders").withSymbol("orderkey", "O"),
                                                node(EnforceSingleRowNode.class,
                                                        node(AggregationNode.class,
                                                                project(
                                                                        project(
                                                                                node(ValuesNode.class)))).withSymbol("min", "MIN").withSymbol("max", "MAX") // withSymbols methods should be called for aggregationNode
)
                                        )))));
```

kokosing replied:
change this for all the queries

AlekseiS replied:
Good point. I'll do it.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -384,6 +384,48 @@ public void testQuantifiedComparisonGreaterEqualAny()
+                                                        tableScan("orders")),
+                                                node(EnforceSingleRowNode.class,
+                                                        node(AggregationNode.class,
+                                                                project(
```
I think you can replace nested project nodes as `anyTree`.

AlekseiS replied:
Done.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -318,10 +319,69 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
+                checkState(outputColumnType.isOrderable(), "Subquery result type must be orderable");
+                FunctionRegistry functionRegistry = metadata.getFunctionRegistry();
+                Symbol subValue = symbolAllocator.newSymbol(aggregationFunction.toString(), outputColumnType);
+                subqueryPlan = new AggregationNode(
```
I think you can add a commit where you extract creation aggregation node with min max aggreagtions for the subquery as separate method. Which could be the used in several rewrites for quantified comparisons.

AlekseiS replied:
I think it's used in 2 places and both of them apply this node differently.

AlekseiS commented on presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4:
```diff
@@ -250,6 +250,7 @@ predicated
 
 predicate[ParserRuleContext value]
     : comparisonOperator right=valueExpression                            #comparison
+    | comparisonOperator comparisonQuantifier '(' query ')'               #quantifiedComparison
```
Regarding the rewrite, this is what my code is doing. In a planner it'll rewrite =ANY to IN() and <>ALL to NOT IN() which are then rewritten/optimized to use join.

kokosing replied:
ok, thanks.

AlekseiS commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/QuantifiedComparisonExpression.java:
```diff
@@ -0,0 +1,103 @@
+        SOME,
+    }
+
+    private final ComparisonExpression.Type comparisonType;
```
Yes, it seems like a good idea. I'll do it.

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -291,6 +292,33 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
             // given subquery is already appended
             return subPlan;
         }
+        switch (quantifiedComparison.getComparisonType()) {
```
I noticed that github indeed displays some commits out of order. I'll make sure I touch all commits for the next upload.

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -291,6 +292,33 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
         }
+        switch (quantifiedComparison.getComparisonType()) {
+            case EQUAL:
+                switch (quantifiedComparison.getQuantifier()) {
```
Well, nested switches with 'break's in between may look a bit weird.
However, I've tried using 'if's with 'isOneOf' as an experiment right now and it seems like it doesn't add much clarity.
I think the main issue impacting readability is that some cases are long. I've extracted 2 of them into their own methods, so overall switch looks better now.

kokosing replied:
thanks

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -291,6 +292,33 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
+                        // A = ANY B <=> A IN B
+                        InPredicate inPredicate = new InPredicate(quantifiedComparison.getValue(), quantifiedComparison.getSubquery());
+                        subPlan.getTranslations().addIntermediateMapping(quantifiedComparison, inPredicate);
+                        return appendInPredicateApplyNode(subPlan, inPredicate, correlationAllowed);
```
Can you please provide more details on your suggestion? If I understood correctly, you suggest rewriting "=ANY" to "IN" in AstBuilder, so that there's no need to deal with it later. If so, I though about this approach before and didn't like it. First of all, AST is close a representation of nodes of an actual query, so doing structural rewrites there seems like not a good idea. Next, the logic will be scattered across AST rewrite and plan generation, unlike now when it's all in a planner.

kokosing replied:
ok. Sounds reasonable. thanks.

AlekseiS commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -317,6 +355,17 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
                         return appendQuantifiedComparisonApplyNode(subPlan, rewrittenAny, correlationAllowed);
+                    case ANY:
+                    case SOME:
+                        // A <> ANY B <=> min B <> max B || A <> min B <=> !(min B = max B && A = min B) <=> !(A = ALL B)
```
Well, that's what it essentially does, because ... <=>!(A = ALL B) <=> !(A IN B).
However, for an implementation I'm reusing (A=ALL B) case, so my equation stops at that point.

AlekseiS replied:
But as an explanation comment, it's probably simpler. I'll change the comment.

kokosing replied:
Yes, you are right. Thanks

AlekseiS commented:
@martint and @kokosing I uploaded a new version addressing your comment. Also, commits on github are ordered correctly now. Please have a look.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java:
```diff
@@ -1007,6 +1008,25 @@ public void testInValidJoinOnClause()
+        analyze("SELECT * FROM t1 WHERE t1.a <= ALL (VALUES 10, 20)");
+        assertFails(MULTIPLE_FIELDS_FROM_SUBQUERY, "SELECT * FROM t1 WHERE t1.a = ANY (SELECT 1, 2)");
+        assertFails(TYPE_MISMATCH, "SELECT * FROM t1 WHERE t1.a = SOME (VALUES ('abc'))");
+        assertFails(MISMATCHED_SET_COLUMN_TYPES, "SELECT * FROM t1 WHERE t1.a = ALL (VALUES 10, 'abc')");
```
Sorry for not being precise. This exactly I want meant. IMO it is useful to make sure that query fails in an expected way and not with NPE for example. However, you check the error code so it I think it is enough. Never mind the comment.

martint replied:
I don't think this last test exercises the quantified comparison code at all. The query will fail when analyzing the expression `VALUES 10, 'abc'`.

AlekseiS commented:
Rebased. @martint please have a look

martint commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -160,7 +160,7 @@
     private final Set<Expression> columnReferences = newIdentityHashSet();
     private final IdentityHashMap<Expression, Type> expressionTypes = new IdentityHashMap<>();
-    private final Set<QuantifiedComparisonExpression> subqueryQuantifiedComparisons = newIdentityHashSet();
+    private final Set<QuantifiedComparisonExpression> quantifiedComparisons = newIdentityHashSet();
```
Use the name "quantifiedComparisons" in the commit that introduce the field. No point in adding it with one name just to rename it in the commit that follows.

AlekseiS replied:
Done.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -265,6 +277,23 @@ private PlanBuilder appendExistSubqueryApplyNode(PlanBuilder subPlan, ExistsPred
+            // given subquery is already appended
+            return subPlan;
+        }
+        throw new SemanticException(NOT_SUPPORTED, quantifiedComparison, "Quantified comparison with the given quantifier and comparison type is not supported");
```
I'd rephrase the error message such that it's obvious to the user what the problem is. Specifically, include the operator and quantifier. I might even go with something like:

`">= ALL (...)" expressions not yet supported`

or something of the sort.

AlekseiS replied:
Rephrased the message

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -291,6 +294,33 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
+                                quantifiedComparison.getSubquery());
+                        Expression notAny = new NotExpression(rewrittenAny);
+                        subPlan.getTranslations().addIntermediateMapping(quantifiedComparison, notAny);
+                        return appendQuantifiedComparisonApplyNode(subPlan, rewrittenAny, correlationAllowed);
```
Is `rewrittenAny` in this line intentional? Or did you mean `notAny`? If it's a mistake, make sure to add a test that reproduces the issue.

AlekseiS replied:
Yes, it's the intended implementation. We know that `A <> ALL B` is equivalent to `NOT (A = ANY B)`, so I add a rewrite for the initial `quantifiedComparison` to `notAny`. However, we still need to plan `ANY B` part, so I call the same function to plan `rewrittenAny`.
Let me know if there's any way I can make it more clear in the code.
Also, current tests, including integration tests, should catch any issues with this code.

martint replied:
Got it. Just add that as a comment above that line.

AlekseiS replied:
Added a comment for both cases.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -320,10 +325,76 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
 
+    private PlanBuilder planQuantifiedOrderable(PlanBuilder subPlan, QuantifiedComparisonExpression quantifiedComparison, boolean correlationAllowed)
+    {
+        checkArgument(quantifiedComparison.getComparisonType() == LESS_THAN || quantifiedComparison.getComparisonType() == LESS_THAN_OR_EQUAL ||
```
You can do `EnumSet.of(LESS_THAN, LESS_THAN_OR_EQUAL, ...).contains(quantifiedComparison.getComparisonType())`

AlekseiS replied:
Good idea. Done.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -320,10 +325,76 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
+                }
+                break;
+        }
+        throw new IllegalStateException("Unexpected quantifier: " + quantifiedComparison.getQuantifier());
```
This should be IllegalArgumentException or UnsupportedOperationException

AlekseiS replied:
Changed to IllegalArgumentException

AlekseiS replied:
Also, removed the test with VALUES 10, 'abc', because it indeed is a bit unrelated.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -316,13 +318,24 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
+                                quantifiedComparison.getSubquery());
+                        Expression notAll = new NotExpression(rewrittenAll);
+                        subPlan.getTranslations().addIntermediateMapping(quantifiedComparison, notAll);
+                        return appendQuantifiedComparisonApplyNode(subPlan, rewrittenAll, correlationAllowed);
```
Similar comment as before... `rewrittenAll` or `notAll` in this line?

AlekseiS replied:
Yes, it's expected. See my reply for `rewrittenAny`.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -332,7 +345,51 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
         }
-        throw new SemanticException(NOT_SUPPORTED, quantifiedComparison, "Quantified comparison with the given quantifier and comparison type is not supported");
+        // all cases are checked, so this exception should never be thrown
+        throw new IllegalStateException("Unexpected combination of quantifier and comparison type: " +
```
From the point of view of this method, this should be either an IllegalArgumentException or UnsupportedOperationException. It could happen if the ComparisonType enum is extended in the future.

AlekseiS replied:
Changed to IllegalArgumentException and rephrased the message

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java:
```diff
@@ -332,7 +345,51 @@ private PlanBuilder appendQuantifiedComparisonApplyNode(PlanBuilder subPlan, Qua
+        List<Expression> outputColumnReference = ImmutableList.of(getOnlyElement(subqueryRelationPlan.getOutputSymbols()).toSymbolReference());
+        Type outputColumnType = getOnlyElement(subqueryRelationPlan.getDescriptor().getAllFields()).getType();
+        if (!outputColumnType.isOrderable()) {
+            throwNotSupportedException(quantifiedComparison, "Quantified comparison with given parameters on unorderable type");
```
"given parameters" is too generic. Try to make the error message more explicit. Also, include the name of the type.

martint replied:
I think it should be possible to implement this without requiring the type to be orderable. 

`A = ALL B <=> A = arbitrary B && count(distinct B) = 1`

We might be able to rewrite the query along these lines:

```
x = (
    SELECT arbitrary(y)
    FROM B
    HAVING count(DISTINCT y) = 1
)
```

Computing `count(DISTINCT y)` could get expensive. We could further improve it by doing something like this, which takes advantage of the distinct-limit optimization.

```
x = (
    SELECT arbitrary(y) 
    FROM (
        SELECT DISTINCT y
        FROM B
        LIMIT 2
    )
    HAVING count(DISTINCT y) = 1
)
```

AlekseiS replied:
Rephrased the message to include quantified comparison and the type.
I was thinking about using 'count distinct' to implement it for unorderable types, but it would be less efficient than min/max. So, I decided to deal only with orderable types in this patch and will create a follow up task to support =ALL/<>ANY with unorderable types.
What do you think?

martint replied:
With the second form above, it shouldn't be that much more expensive. The Distinct-limit optimization will cause the subquery to bail out as soon as it sees two distinct values, and won't use more memory than necessary to store just the two values.

It's fine to leave it as a future enhancement, though. Please file an issue to track it.

AlekseiS replied:
Yes, distinct with limit is a good idea. I'll file a task with these details once this code is merged.

AlekseiS commented:
@martint I addressed the comments and rebased. Please have a look again.

AlekseiS commented:
After I fixed a few cases of ComparisonExpression.Type added by new commits build and tests pass. I'll merge the change. @martint thanks for the review!

yuzisun commented:
Is there any plan to push down ANY/ALL/SOME predicates ?

kokosing commented:
@yuzisun What do you mean? ANY/ALL/SOME are rewritten to other predicates, that means they won't be pushed down in that form.

yuzisun commented:
Is ANY gets rewritten to IN and then turns into a series of ORs ? When we test IN clause with a large set of values (50k elements) which is our use case, the system falls apart, it took 20s+ to just parse the SQL statement. So we implemented our own ANY operator according to the framework and pushed down the values to connector level.
https://github.com/bloomberg/presto/commit/4a5279e6d773cce0c3d9883c6a779781ec6077f3

kokosing commented:
ANY accepts subquery, it is the rewritten to some form of (semi) join, depending on comparison type. IN also can accept subquery, I would  prefer that form instead of hard coding 50k values. AFAIR sql query can be at most 1MB, so you are close to the limit.

As I understand your ANY operator is similar to `array_postition()` function. Is that ANSI SQL compliant? Is there PR with that?

adamjshook commented:
@kokosing I had opened https://github.com/prestodb/presto/pull/5275 with our implementation but it was ultimately deemed the wrong way to implement it (which is fine).  The 50k elements are deterministic (as far as I know) but we have no table containing these values to use as a subquery.  Maybe that would be a better approach overall -- something to look into.
